modules/order/commerce_order.module

```diff
@@ -181,3 +181,41 @@ function commerce_order_mail($key, &$message, $params) {
   $message['subject'] = $params['subject'];
   $message['body'][] = $params['body'];
 }
+
+/**
+ * Implements hook_ENTITY_TYPE_predelete() for user entity.
+ */
+function commerce_order_user_predelete(EntityInterface $entity) {
+  /** @var \Drupal\commerce_order\Entity\OrderInterface[] $list */
+  $list = \Drupal::entityTypeManager()
+    ->getStorage('commerce_order')
+    ->loadByProperties([
+      'uid' => $entity->id(),
+      'state' => 'draft',
+    ]);
+
+  foreach ($list as $commerce_order) {
+    $commerce_order->delete();
+  }
+}
+
+/**
+ * Implements hook_user_cancel().
+ */
+function commerce_order_user_cancel($edit, $account, $method) {
+  switch ($method) {
+    case 'user_cancel_reassign':
```

Comment:
Wouldn't we want to also run this at `user_cancel_delete` method, too? So that orders don't have a broken reference.